src/sage/modules/free_module_pseudomorphism.py

```diff
@@ -0,0 +1,224 @@
+
+class FreeModulePseudoMorphism(Morphism):
+    r"""
+    Let `M, M'` be free modules over a ring `R`, `\theta: R \to R` a ring
+    homomorphism, and `\theta: R \to R` a derivation i.e. an additive
+    map such that
+
+    `\delta(xy) = x\delta(y) + \delta(x)y`.
+
+    Then a pseudomorphism `f : M to M` is a map such that
+
+    `f(x + y) = f(x) + f(y)`
+    `f(\lambda x) = `\theta(\lambda)f(x) + \delta(\lambda)x`
+
+    The pair `(\theta, \delta)` may be referred to as the *twist* of
+    the morphism.
+
+    TESTS::
+
+        sage: V = ZZ^2
+        sage: f = V.pseudohom([V.1, -2*V.0]); f
+        Free module pseudomorphism defined by the matrix
+        [ 0  1]
+        [-2  0]
+        Domain: Ambient free module of rank 2 over the principal ideal domain Integer Ring
+        Codomain: Ambient free module of rank 2 over the principal ideal domain Integer Ring
+        sage: f(V((1, 2)))
+        (-4, 1)
+
+        sage: P.<x> = ZZ[]; deriv = P.derivation()
+        sage: M = P^2
+        sage: f = M.pseudohom([[1, 2*x], [x, 1]], deriv, side="right"); f
+        Free module pseudomorphism defined as left-multiplication by the matrix
+        [  1 2*x]
+        [  x   1]
+        twisted by the derivation d/dx
+        Domain: Ambient free module of rank 2 over the integral domain Univariate Polynomial Ring in x over Integer Ring
+        Codomain: Ambient free module of rank 2 over the integral domain Univariate Polynomial Ring in x over Integer Ring
+        sage: e = M((2*x^2 + 3*x + 1, x^3 + 7*x + 4))
+        sage: f(e)
+        (2*x^4 + 16*x^2 + 15*x + 4, 3*x^3 + 6*x^2 + 8*x + 11)
+        sage: f = M.pseudohom([[1, 2], [1, 1]], deriv)
+        sage: f(e)
+        (x^3 + 2*x^2 + 14*x + 8, x^3 + 7*x^2 + 13*x + 13)
+    """
+    def __init__(self, domain, base_morphism, twist=None, codomain=None, side="left"):
+        """
+        Constructs a pseudomorphism of free modules.
+
+        INPUT:
+            -  ``domain``   - the domain of the pseudomorphism; a free module 
+
+            -  ``base_morphism`` - either a morphism or a matrix defining a
+                                   morphism
+
+            -  ``twist`` - a twisting morphism, this is either a morphism or 
+                           a derivation (default: None)
+
+            -  ``codomain`` - the codomain of the pseudomorphism; a free
+                              module  (default: None)
+
+            - side -- side of the vectors acted on by the matrix 
+                      (default: ``"left"``)
+
+        EXAMPLES::
+
+            sage: F = GF(25); V = F^3; twist = F.frobenius_endomorphism(5)
+            sage: phi = V.pseudohom(matrix(F,3,[1..9]), twist)
+            sage: type(phi)
+            <class 'sage.modules.free_module_pseudomorphism.FreeModulePseudoMorphism'>
+        """
+        from sage.structure.element import is_Matrix
+        if is_Matrix(base_morphism):
+            self.base_morphism = domain.hom(base_morphism, codomain)
```

Comment:
Should be better to write `self._base_morphism`; the same for the other private parameters.

Reply:
Also why don't you work directly with the matrix (instead of building this linear morphism, which is not really relevant IMO).